lib/AST/CompilerPlugin.cpp

```diff
@@ -52,183 +50,43 @@ struct MetadataArrayRef {
-#if !defined(_WIN32)
-      errorMsg = dlerror();
-#endif
-      Diags.diagnose(SourceLoc(), diag::compiler_plugin_not_loaded, path,
-                     errorMsg);
-      continue;
-    }
-    auto moduleName = llvm::sys::path::filename(path);
-#if !defined(_WIN32)
-    moduleName.consume_front("lib");
-#endif
-    moduleName.consume_back(LTDL_SHLIB_EXT);
-    auto *getter = getMacroRegistrationPropertyGetter(lib, moduleName, *this);
-    if (!getter) {
-      Diags.diagnose(SourceLoc(),
-                     diag::compiler_plugin_missing_macro_declaration,
-                     moduleName, path, ALLMACROS_PROPERTY_NAME);
-      continue;
-    }
-    // Note: We don't currently have a way to poke at the contents of a Swift
-    // array `[Any.Type]` from C++. But this should not be an issue for release
-    // toolchains where user-defined macros will be used.
-#if SWIFT_SWIFT_PARSER
-    const void *const *metatypesAddress;
-    ptrdiff_t metatypeCount;
-    swift_ASTGen_getMacroTypes(getter, &metatypesAddress, &metatypeCount);
-    ArrayRef<const void *> metatypes(metatypesAddress, metatypeCount);
-    for (const void *metatype : metatypes) {
-      auto plugin = new CompilerPlugin(metatype, lib, *this);
-      auto name = plugin->getName();
-      addLoadedPlugin(name, plugin);
-    }
-    free(const_cast<void *>((const void *)metatypes.data()));
-#endif // SWIFT_SWIFT_PARSER
-  }
-}
-
 using WitnessTableLookupFn = const void *(const void *type,
                                           const void *protocol);
 
 #if SWIFT_SWIFT_PARSER
 extern "C" WitnessTableLookupFn swift_conformsToProtocol;
 #endif
 
-CompilerPlugin::CompilerPlugin(const void *metadata, void *parentLibrary,
-                               ASTContext &ctx)
-   : metadata(metadata), parentLibrary(parentLibrary)
+CompilerPlugin::CompilerPlugin(
+    const void *metadata, void *parentLibrary, const void *witnessTable)
+   : metadata(metadata), parentLibrary(parentLibrary),
+     witnessTable(witnessTable)
 {
+  assert(witnessTable && "Type does not conform to _CompilerPlugin");
+  kind = invokeKind();
+}
+
+CompilerPlugin::~CompilerPlugin() { }
+
+CompilerPlugin *CompilerPlugin::fromMetatype(
+    const void *metadata, ASTContext &ctx
+) {
 #if !SWIFT_SWIFT_PARSER
   auto *swift_conformsToProtocol = reinterpret_cast<WitnessTableLookupFn *>(
       ctx.getAddressOfSymbol("swift_conformsToProtocol"));
 #endif
   void *protocolDescriptor =
       ctx.getAddressOfSymbol(COMPILER_PLUGIN_PROTOCOL_DESCRIPTOR);
-  assert(swift_conformsToProtocol);
-  assert(protocolDescriptor);
-  witnessTable = swift_conformsToProtocol(metadata, protocolDescriptor);
-  assert(witnessTable && "Type does not conform to _CompilerPlugin");
-  auto returnedName = invokeName();
-  name = ctx.getIdentifier(returnedName).str();
-  free(const_cast<void *>((const void *)returnedName.data()));
-  kind = invokeKind();
-}
+  if ((&swift_conformsToProtocol == nullptr) || !protocolDescriptor)
```

Comment:
Isn't `&swift_conformsToProtocol` never null since it's either a local variable or an extern declaration? `swift_conformsToProtocol` could be null.

Reply:
Ah... yeah, I got sloppy here. I'll fix it in a follow-up.